osquery/tables/system/linux/md_tables.cpp

```diff
@@ -0,0 +1,733 @@
+ *
+ * @param  state state field of mdu_array_info_t
+ *
+ * @return stringified state
+ */
+std::string getSuperBlkStateStr(int state) {
+  if (state == 0)
+    return "unknown";
+
+  std::string s;
+
+#ifdef MD_SB_CLEAN
+  if ((1 << MD_SB_CLEAN) & state)
+    s += "clean ";
+#endif
+
+#ifdef MD_SB_ERRORS
+  if ((1 << MD_SB_ERRORS) & state)
+    s += "errors ";
+#endif
+
+#ifdef MD_SB_BBM_ERRORS
+  if ((1 << MD_SB_BBM_ERRORS) & state)
+    s += "bbm_errors ";
+#endif
+
+#ifdef MD_SB_BLOCK_CONTAINER_RESHAPE
+  if ((1 << MD_SB_BLOCK_CONTAINER_RESHAPE) & 1)
+    s += "container_reshape ";
+#endif
+
+#ifdef MD_SB_BLOCK_VOLUME
+  if ((1 << MD_SB_BLOCK_VOLUME) & 1)
+    s += "block_activation ";
+#endif
+
+#ifdef MD_SB_CLUSTERED
+  if ((1 << MD_SB_CLUSTERED) & 1)
+    s += "clustered ";
+#endif
+
+#ifdef MD_SB_BITMAP_PRESENT
+  if ((1 << MD_SB_BITMAP_PRESENT) & 1)
+    s += "bitmap_present ";
+#endif
+
+  boost::algorithm::trim(s);
+  return s;
+}
+
+// For use with unique_ptr of file close as a hacky way of preventing fd leaks
+auto fClose = [](int* fd) { close(*fd); };
+
+bool MD::getDiskInfo(std::string arrayName, mdu_disk_info_t& diskInfo) {
+  std::map<std::string, std::string> results;
+  int fd;
+
+  std::unique_ptr<int, decltype(fClose)> _(
+      &(fd = open(arrayName.c_str(), O_RDONLY)), fClose);
+  int status = ioctl(fd, GET_DISK_INFO, &diskInfo);
+
+  if (status == -1) {
+    LOG(WARNING) << "Call to ioctl 'GET_DISK_INFO' " << arrayName
+                 << " failed: " << strerror(errno);
+    return false;
+  }
+
+  return true;
+}
+
+bool MD::getArrayInfo(std::string name, mdu_array_info_t& array) {
+  std::map<std::string, std::string> results;
+  int fd;
+
+  std::unique_ptr<int, decltype(fClose)> _(&(fd = open(name.c_str(), O_RDONLY)),
+                                           fClose);
+  int status = ioctl(fd, GET_ARRAY_INFO, &array);
+
+  if (status == -1) {
+    LOG(ERROR) << "Call to ioctl 'GET_ARRAY_INFO' for " << name
+               << " failed: " << strerror(errno);
+    return false;
+  }
+
+  return true;
+}
+
+inline void parseMDPersonalities(std::string& line,
+                                 std::vector<std::string>& result) {
+  std::vector<std::string> enabledPersonalities = split(line, " ");
+  for (auto& p : enabledPersonalities) {
+    boost::algorithm::trim(p);
+    result.push_back(p.substr(1, p.length() - 2));
+  }
+}
+
+void parseMDAction(std::string& line, MDAction& result) {
+  /* Make assumption that recovery/resync format is [d+]% ([d+]/[d+])
+   * finish=<duration> speed=<rate> */
+  std::vector<std::string> pieces(split(line, " "));
+  if (pieces.size() != 4) {
+    LOG(WARNING) << "Unexpected recovery/resync line format: " << line;
+    return;
+  }
+  trimStrs(pieces);
+
+  result.progress = pieces[0] + " " + pieces[1];
+
+  std::size_t start = pieces[2].find_first_not_of("finish=");
+  if (start != std::string::npos) {
+    result.finish = pieces[2].substr(start);
```

Comment:
I don' t think we need to check the size in this context since we're already checking if start is `npos`?